src/rsp_restspawner/spawner.py

```diff
@@ -35,6 +39,63 @@ class LabStatus(str, Enum):
     FAILED = "failed"
 
 
+@dataclass(frozen=True, slots=True)
+class SpawnEvent:
+    """JupyterHub spawning event."""
+
+    progress: int
+    """Percentage of progress, from 0 to 100."""
+
+    message: str
+    """Event description."""
+
+    severity: str
+    """Log message severity."""
+
+    complete: bool = False
+    """Whether the event indicated spawning is done."""
+
+    failed: bool = False
+    """Whether the event indicated spawning failed."""
+
+    @classmethod
+    def from_sse(cls, sse: ServerSentEvent, progress: int) -> SpawnEvent:
+        """Convert from a server-sent event from the lab controller.
+
+        Parameters
+        ----------
+        sse
+            Event from the lab controller.
+        progress
+            Current progress percentage. Parsing of the progress events that
+            communicate this must be done outside of this class.
+        """
+        if sse.event == "complete":
+            return cls(
+                progress=90, message=sse.data, severity="info", complete=True
+            )
+        elif sse.event == "info":
```

Comment:
I would probably have done something silly with functools.partial and then updated the instance object thus created by changing progress or severity if needed, and adding complete/failed if needed, but this is clearer.